### DISCUSSION TITLE
How to create a branch on remote and push to it?

### DISCUSSION THREAD

**Discussion by MMachado05** [jesseduffield/lazygit#3865] [Q&A] [answered]
Say I'm on `main` and want to create a new branch to be tracked on remote, we'll call it `branch` for brevity. In the CLI, I would `git checkout -b branch`, edit, add, and commit, until finally doing a `git push -u origin branch`. That makes a new branch on remote, and then my local `branch` branch will be pushing to it whenever I push.

I can't seem to figure this out on Lazygit. I'll make a branch from my local `main` branch, `u` for upstream options, `s` to set upstream, but it only lets me use pre-existing remote branches, rather than being able to push to a new one. Am I misunderstanding something here?

**Comment by stefanhaller** [comment, marked as answer]
Create your branch locally, e.g. from your local main. Make your commits. Once you're ready to push it to the remote, hit shift-P; it will ask you to which remote you want to push, and with which remote branch name.

**Reply by MMachado05** [reply]
Thank you! I feel a bit silly now seeing how obvious it was... 😅 But your concise answer is greatly appreciated!

**Reply by canton7** [reply]
Sorry to resurrect this, but that doesn't seem to work. I get the following (sorry for the screenshot, I can't work out how to copy text in a nice way):

![image](https://github.com/user-attachments/assets/29c38b86-dbde-4084-9efc-52f105107bab)

I want to do `git push -u origin HEAD` or similar.

@stefanhaller Do you have any ideas?

**Reply by suside** [reply]
If you already have upstream branch like `origin/main` focus on the "Local branches" and then press `uu` (first `u` for `Upstream options`, second for `Unset upstream of selected branch`). Now when you do push lazygit will ask for a name of your new upstream branch.

**Comment by Jnolcox** [comment]
Bumping this as I am getting the same issue - I can't push a new remote branch

**Reply by mark2185** [reply]
When you say "can't", what is the exact error message?

**Reply by Jnolcox** [reply]
The same error message as the above.

**Reply by mark2185** [reply]
So, you create a new branch with `n`, press`shift+P`, and the error pops up?

If not, please list the steps you took.

**Reply by Jnolcox** [reply]
Correct. 

The branch does not yet exist remotely when executing these steps. After pushing, the error pops up.

**Reply by mark2185** [reply]
Does the branch have an upstream set? What if you unset it through the menu under `u`?

**Reply by Jnolcox** [reply]
It does not have an upstream set. 

`The selected branch has no upstream (or the upstream is not stored locally)`

**Reply by canton7** [reply]
In my case, the branch does have an upstream of `origin/main`, because it was created with `git checkout -b branch_name origin/main`.

Unsetting the upstream works.

**Reply by Jnolcox** [reply]
Correct me if I'm wrong, but when exercising the option to create a new branch from within LazyGit, it should create, checkout and set upstream. It seems it is only creating it locally but drops the ball on push.
